File: windows/ReviewAnything/Services/MarkdownParser.cs
using System.Security.Cryptography;
using System.Text;

namespace ReviewAnything.Services;

public record ParsedChunk(string Content, string HeadingPath);

public static class MarkdownParser
{
    public static List<ParsedChunk> Parse(string content)
    {
        var lines = content.Split('\n');
        var chunks = new List<ParsedChunk>();
        var headingStack = new List<string>();
        var currentContent = new StringBuilder();

        void FlushChunk()
        {
            var text = currentContent.ToString().Trim();
            if (text.Length >= 20)
            {
                chunks.Add(new ParsedChunk(text, string.Join("/", headingStack)));
            }
            currentContent.Clear();
        }

        foreach (var line in lines)
        {
            var trimmed = line.Trim();
            if (trimmed.StartsWith("#"))
            {
                FlushChunk();
                var level = trimmed.TakeWhile(c => c == '#').Count();
                var title = trimmed[level..].Trim();
                while (headingStack.Count >= level)
                {
                    headingStack.RemoveAt(headingStack.Count - 1);
                }
                headingStack.Add(title);
            }
            else
            {
                currentContent.AppendLine(line);
            }
        }
        FlushChunk();

        return chunks.Count > 0 ? chunks : new List<ParsedChunk> { new(content.Trim(), "") };
    }

    public static string ComputeHash(string content)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(content));
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }
}

File: android/app/src/main/java/com/reviewanything/app/service/MarkdownParser.kt
package com.reviewanything.app.service

import java.security.MessageDigest

data class ParsedChunk(
    val content: String,
    val headingPath: String
)

object MarkdownParser {

    fun parse(content: String): List<ParsedChunk> {
        val lines = content.lines()
        val chunks = mutableListOf<ParsedChunk>()
        val headingStack = mutableListOf<String>()
        val currentContent = StringBuilder()

        fun flushChunk() {
            if (currentContent.isNotBlank()) {
                val text = currentContent.toString().trim()
                if (text.length >= 20) {
                    chunks.add(ParsedChunk(text, headingStack.joinToString("/")))
                }
            }
            currentContent.clear()
        }

        for (line in lines) {
            val trimmed = line.trim()
            if (trimmed.startsWith("#")) {
                flushChunk()
                val level = trimmed.takeWhile { it == '#' }.length
                val title = trimmed.drop(level).trim()
                while (headingStack.size >= level) {
                    headingStack.removeAt(headingStack.size - 1)
                }
                headingStack.add(title)
            } else {
                currentContent.appendLine(line)
            }
        }
        flushChunk()

        return chunks.ifEmpty {
            listOf(ParsedChunk(content.trim(), ""))
        }
    }

    fun computeHash(content: String): String {
        val bytes = MessageDigest.getInstance("SHA-256").digest(content.toByteArray())
        return bytes.joinToString("") { "%02x".format(it) }
    }
}
